### ISSUE TITLE
Triplet loss resulting in 0.2% mAP

### ISSUE THREAD

**Issue by lennartmoritz** [KaiyangZhou/deep-person-reid#357]
I've added the dataset im working with (PRAI-1581) and want to use tripplet loss for training but the results make no sense.
I've already used softmax loss with OSNET and achieved 47,7% mAP and I've used ResNet50 with softmax which achieved 36,4% mAP. Therefore I can presume that the problem is not a faulty implementation of my dataset.

But when i run the following config with triplet loss, I will only get 0.2% mAP and I hope you can help me find out how I am using torchreid wrong. 
I will attach an image to show that by training my model it actually just gets worse over time:
![Anmerkung 2020-07-23 175111](https://user-images.githubusercontent.com/51822945/88308267-24916a00-cd0d-11ea-95c3-6348e89afd77.png)

Also after about the 20th epoch the loss value just stays around 0.3 which reminds me of the cfg.loss.triplet.margin = 0.3 value. Does that make sense?
 
```
Show configuration
adam:
  beta1: 0.9
  beta2: 0.999
cuhk03:
  classic_split: False
  labeled_images: False
  use_metric_cuhk03: False
data:
  combineall: False
  height: 256
  load_train_targets: False
  norm_mean: [0.485, 0.456, 0.406]
  norm_std: [0.229, 0.224, 0.225]
  root: /net/merkur/storage/deeplearning/datasets/reid
  save_dir: /net/merkur/storage/deeplearning/users/morlen/log/resnet50_prai1581_triplet_cosinelr6
  sources: ['prai1581']
  split_id: 0
  targets: ['prai1581']
  transforms: ['random_flip', 'random_erase', 'random_crop']
  type: image
  width: 128
  workers: 4
loss:
  name: triplet
  softmax:
    label_smooth: True
  triplet:
    margin: 0.3
    weight_t: 1.0
    weight_x: 0.0
market1501:
  use_500k_distractors: False
model:
  load_weights: 
  name: resnet50
  pretrained: True
  resume: 
rmsprop:
  alpha: 0.99
sampler:
  num_instances: 4
  train_sampler: RandomIdentitySampler
  train_sampler_t: RandomIdentitySampler
sgd:
  dampening: 0.0
  momentum: 0.9
  nesterov: False
test:
  batch_size: 450
  dist_metric: euclidean
  eval_freq: 20
  evaluate: False
  normalize_feature: True
  ranks: [1, 5, 10, 20]
  rerank: False
  start_eval: 0
  visrank: False
  visrank_topk: 10
train:
  base_lr_mult: 0.1
  batch_size: 104
  fixbase_epoch: 10
  gamma: 0.1
  lr: 0.002
  lr_scheduler: cosine
  max_epoch: 125
  new_layers: ['classifier']
  open_layers: ['classifier']
  optim: amsgrad
  print_freq: 20
  seed: 1
  staged_lr: False
  start_epoch: 0
  stepsize: [20]
  weight_decay: 0.0005
use_gpu: True
video:
  pooling_method: avg
  sample_method: evenly
  seq_len: 15
```
I will add some output here too:
```
=> Start training
* Only train ['classifier'] (epoch: 1/10)
epoch: [1/125][20/176]	time 0.148 (0.313)	data 0.000 (0.092)	eta 1:54:45	loss_t 5.6625 (6.1717)	loss_x 6.6618 (6.6707)	acc 0.0000 (0.0962)	lr 0.002000
epoch: [1/125][40/176]	time 0.136 (0.226)	data 0.000 (0.046)	eta 1:22:35	loss_t 6.0268 (6.1738)	loss_x 6.6604 (6.6653)	acc 0.0000 (0.0721)	lr 0.002000
epoch: [1/125][60/176]	time 0.137 (0.197)	data 0.000 (0.031)	eta 1:12:03	loss_t 6.0984 (6.1199)	loss_x 6.6612 (6.6637)	acc 0.0000 (0.0641)	lr 0.002000
epoch: [1/125][80/176]	time 0.136 (0.183)	data 0.000 (0.023)	eta 1:06:50	loss_t 5.5100 (6.1143)	loss_x 6.6605 (6.6629)	acc 3.8462 (0.1082)	lr 0.002000
epoch: [1/125][100/176]	time 0.140 (0.174)	data 0.000 (0.019)	eta 1:03:41	loss_t 6.0706 (6.0967)	loss_x 6.6606 (6.6624)	acc 0.0000 (0.1250)	lr 0.002000
epoch: [1/125][120/176]	time 0.139 (0.170)	data 0.000 (0.015)	eta 1:01:54	loss_t 6.3942 (6.1170)	loss_x 6.6606 (6.6621)	acc 0.0000 (0.1522)	lr 0.002000
epoch: [1/125][140/176]	time 0.166 (0.166)	data 0.000 (0.013)	eta 1:00:31	loss_t 6.7135 (6.0950)	loss_x 6.6605 (6.6619)	acc 0.0000 (0.1717)	lr 0.002000
epoch: [1/125][160/176]	time 0.153 (0.164)	data 0.000 (0.012)	eta 0:59:35	loss_t 5.6027 (6.1078)	loss_x 6.6606 (6.6617)	acc 0.0000 (0.1562)	lr 0.002000
...
epoch: [16/125][20/176]	time 0.448 (0.489)	data 0.000 (0.027)	eta 2:37:42	loss_t 0.3140 (0.3217)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0000)	lr 0.001930
epoch: [16/125][40/176]	time 0.437 (0.466)	data 0.000 (0.013)	eta 2:30:01	loss_t 0.3143 (0.3204)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0000)	lr 0.001930
epoch: [16/125][60/176]	time 0.465 (0.463)	data 0.000 (0.009)	eta 2:28:52	loss_t 0.3013 (0.3152)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0000)	lr 0.001930
epoch: [16/125][80/176]	time 0.449 (0.462)	data 0.000 (0.007)	eta 2:28:23	loss_t 0.3251 (0.3131)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0120)	lr 0.001930
epoch: [16/125][100/176]	time 0.451 (0.459)	data 0.000 (0.005)	eta 2:27:25	loss_t 0.3010 (0.3118)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0865)	lr 0.001930
epoch: [16/125][120/176]	time 0.451 (0.457)	data 0.000 (0.005)	eta 2:26:37	loss_t 0.3256 (0.3108)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0721)	lr 0.001930
epoch: [16/125][140/176]	time 0.437 (0.454)	data 0.000 (0.004)	eta 2:25:32	loss_t 0.3028 (0.3102)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0687)	lr 0.001930
epoch: [16/125][160/176]	time 0.441 (0.452)	data 0.000 (0.003)	eta 2:24:43	loss_t 0.3011 (0.3095)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0962)	lr 0.001930
...
epoch: [60/125][20/176]	time 0.447 (0.484)	data 0.000 (0.028)	eta 1:33:28	loss_t 0.3005 (0.3013)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0000)	lr 0.001088
epoch: [60/125][40/176]	time 0.441 (0.466)	data 0.000 (0.014)	eta 1:29:58	loss_t 0.3003 (0.3009)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0962)	lr 0.001088
epoch: [60/125][60/176]	time 0.442 (0.460)	data 0.000 (0.009)	eta 1:28:39	loss_t 0.3003 (0.3008)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.1923)	lr 0.001088
epoch: [60/125][80/176]	time 0.450 (0.458)	data 0.000 (0.007)	eta 1:28:03	loss_t 0.3007 (0.3007)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.2404)	lr 0.001088
epoch: [60/125][100/176]	time 0.450 (0.457)	data 0.000 (0.006)	eta 1:27:41	loss_t 0.3001 (0.3006)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.2981)	lr 0.001088
epoch: [60/125][120/176]	time 0.451 (0.456)	data 0.000 (0.005)	eta 1:27:21	loss_t 0.3000 (0.3006)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.2804)	lr 0.001088
epoch: [60/125][140/176]	time 0.453 (0.455)	data 0.000 (0.004)	eta 1:27:03	loss_t 0.3004 (0.3005)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.2679)	lr 0.001088
epoch: [60/125][160/176]	time 0.451 (0.455)	data 0.000 (0.004)	eta 1:26:49	loss_t 0.3005 (0.3005)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.2344)	lr 0.001088
...
epoch: [125/125][20/176]	time 0.439 (0.479)	data 0.000 (0.029)	eta 0:01:14	loss_t 0.3000 (0.3000)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0000)	lr 0.000000
epoch: [125/125][40/176]	time 0.450 (0.468)	data 0.000 (0.014)	eta 0:01:03	loss_t 0.3000 (0.3000)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0000)	lr 0.000000
epoch: [125/125][60/176]	time 0.453 (0.460)	data 0.000 (0.010)	eta 0:00:53	loss_t 0.3000 (0.3000)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0641)	lr 0.000000
epoch: [125/125][80/176]	time 0.449 (0.457)	data 0.000 (0.007)	eta 0:00:43	loss_t 0.3000 (0.3000)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.0481)	lr 0.000000
epoch: [125/125][100/176]	time 0.454 (0.456)	data 0.000 (0.006)	eta 0:00:34	loss_t 0.3000 (0.3000)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.1154)	lr 0.000000
epoch: [125/125][120/176]	time 0.436 (0.453)	data 0.000 (0.005)	eta 0:00:25	loss_t 0.3000 (0.3000)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.1923)	lr 0.000000
epoch: [125/125][140/176]	time 0.453 (0.451)	data 0.000 (0.004)	eta 0:00:16	loss_t 0.3000 (0.3000)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.1923)	lr 0.000000
epoch: [125/125][160/176]	time 0.450 (0.452)	data 0.000 (0.004)	eta 0:00:07	loss_t 0.3000 (0.3000)	loss_x 6.6606 (6.6606)	acc 0.0000 (0.1923)	lr 0.000000
=> Final test
##### Evaluating prai1581 (source) #####
Extracting features from query set ...
Done, obtained 4680-by-2048 matrix
Extracting features from gallery set ...
Done, obtained 15258-by-2048 matrix
Speed: 0.0128 sec/batch
Normalzing features with L2 norm ...
Computing distance matrix with metric=euclidean ...
Computing CMC and mAP ...
** Results **
mAP: 0.2%
CMC curve
Rank-1  : 0.1%
Rank-5  : 0.6%
Rank-10 : 0.9%
Rank-20 : 1.7%
Checkpoint saved to "/net/merkur/storage/deeplearning/users/morlen/log/resnet50_prai1581_triplet_cosinelr6/model/model.pth.tar-125"
Elapsed 2:39:56
```
Please help me understand how to use triplet loss with torchreid.

**Comment by KaiyangZhou** [comment]
hmm, interesting

just to confirm, when you used the config you showed above, but change `loss.name` from `triplet` to `softmax`, also use `RandomSampler` instead of the identity-related one, you could get a reasonable performance? if so, it might well indicate triplet loss doesn't work for this dataset, ... but this is indeed too weird to me, tbh

you can try this
```
loss:
  name: triplet
  softmax:
    label_smooth: True
  triplet:
    margin: 0.3
    weight_t: 0.5
    weight_x: 1.0
```
basically this means combining triplet with softmax (just to confirm the problem is indeed from the triplet loss)

btw, your config file looks fine to me

**Comment by lennartmoritz** [comment]
I've done some further testing and it seemed like lowering the learning rate to
`lr: 0.0002`
was able to improve my mAP results to 5.7% after 15/24 epochs (thats when I had to interrupt that training)
Also removing **random_erase** 
`transforms: ['random_flip', 'random_crop']`
from the transforms yielded a minor mAP improvement of 5.03% after 125/125 epochs.

On the other hand by using your proposed combined triplet and cross corellation (softmax) loss
```
loss:
  name: triplet
  softmax:
    label_smooth: True
  triplet:
    margin: 0.3
    weight_t: 0.5
    weight_x: 1.0
```
combined with
```
lr: 0.0015
transforms: ['random_flip', 'random_crop']
```
resulted in 40.9% mAP for resnet50 after 125/125 epochs. Which is the best mAP for resnet50 I was able to achieve so far.
The paper regarding the PRAI1581 dataset reports achieving 39.58% mAP by using their proposed subspace pooling method while using Triplet loss and 37.88% by using identity loss (softmax).
Which could make sense if by "Triplet loss" they actually mean combined Triplet and Cross Correllation Loss. 

Do you think people may actually mean combined loss when they talk about Triplet Loss, as it makes more sense to use combined loss than triplet on its own?

**Comment by KaiyangZhou** [comment]
no, so if you see a paper says "we use triplet loss", it really means they use triplet loss only (unless they're dishonest about the implementation, but that's rare)

otherwise you would see sth like "we combine triplet loss with cross entropy ..."

**Comment by bryanlinnan** [comment]
hi len, i have met same issue as yours, i used triplet loss to train my dataset and the result is not good, did you figure out why?@lennartmoritz

**Comment by KaiyangZhou** [comment]
@lennartmoritz @bryanlinnan so which paper are you guys referring to? can you give a paper link here?

**Comment by lennartmoritz** [comment]
@KaiyangZhou  The paper i was referring to is "[Person Re-identification in Aerial Imagery](https://arxiv.org/abs/1908.05024)" they also have a [GitHub page](https://github.com/stormyoung/PRAI-1581). Then my training parameters may just be unsuited for triplet loss at the moment.

@bryanlinnan Hey, around the time that I first created this issue I searched the web for possible sollutions. Similar problems with triplet loss have been described in issues from other GitHub repositorys but I didn't save the link. 
It was suggested that bad performance with triplet loss may be caused by too high learning rates and that lowering the learning rate in such cases should help. But I have not tried to use only triplet loss anymore, so I can not verify the solution. Using it in combination with cross entropy loss worked fine though.

A possible reason for the loss value getting "stuck" may be that with too high learning rates the net could accidentally learn to decrease its losses by assigning every input image the same feature representation. If that hapens the triplet loss function would calculate the loss be equal to the margin. 

With a = anchor, n = negative, p = positive:
![grafik](https://user-images.githubusercontent.com/51822945/90264308-30c2a000-de51-11ea-85e6-988ebaed1bed.png)
Screenshot from here [here ](https://omoindrot.github.io/triplet-loss) 
 If you find out more about this problem please let me know.

**Comment by danishmansari** [comment]
@KaiyangZhou  @lennartmoritz Hey why do you think the loss is not going below 0.3 is it supposed to be not going below the margin? I think margin is for the minimum separation we want in the feature vectors of positive and negative samples, but why the loss is not going below that?

**Comment by zhtn99** [comment]
I am also reproducing this paper recently. Have you solved these problems thoroughly and succeeded in reproducing them? @lennartmoritz

**Comment by lennartmoritz** [comment]
@zhtn99 i was not able to reproduce their results using triplet loss only. There were no new findings on this issue, so I gave up on it.

But with the combined triplet and cross corellation (softmax) loss as described in an earlier comment, i was able to get more reasonable results. My best result was 49.1% mAP with OSNet for PRAI-1581 with combined loss, which was still slightly worse then my best results with softmax loss.

**Comment by zhtn99** [comment]
Thank you for your reply. Have you reproduced the subspace pooling algorithm proposed in the paper 《Person Re-identification in Aerial Imagery》? Currently, I want to implement the subspace pooling algorithm based on torchreid framework, but the results of 37.88% and 39.58% in the paper are not obtained, and I do not know the problem for the time being. @lennartmoritz
